File: src/go/crc32.go
package main

import	"fmt"
import	"os"

	const sep            = " "
	const poly    uint32 = 0xEDB88320

	var code   int
	var table  = make([]uint32, 256)

func join(a []string) string {
	switch len(a) {
	case 0:
		return ""
	case 1:
		return a[0]
	case 2:
		return a[0] + sep + a[1]
	case 3:
		return a[0] + sep + a[1] + sep + a[2]
	}
	n := len(sep) * (len(a) - 1)
	for i := 0; i < len(a); i++ {
		n += len(a[i])
	}

	b := make([]byte, n)
	bp := copy(b, a[0])
	for _, s := range a[1:] {
		bp += copy(b[bp:], sep)
		bp += copy(b[bp:], s)
	}
	return string(b)
}

func init() {
	for i := 0; i < 256; i++ {
		c := uint32(i)
		for j := 0; j < 8; j++ {
			b := c&1 == 1
			c >>= 1
			if b {
				c ^= poly
			}
		}
		table[i] = c
	}
}

func crc32(s string) uint32 {
	crc := ^uint32(0)
	for i := 0; i < len(s); i++ {
		crc = table[byte(crc)^s[i]] ^ (crc >> 8)
	}
	return ^crc
}

func main() {
        if len(os.Args) != 2 {
	  fmt.Printf("%s [STRING or WORD] ...\n", os.Args[0])
	  code = 1
        } else {
	  fmt.Printf("%0x\n", crc32(os.Args[1]))
        }
	os.Exit(code)
}

File: src/c/crc32.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <inttypes.h>

uint32_t crc32(const char *s) {
	static uint32_t c = 0, crc = 0xffffffff, table[256];

	for (int i = 0; i < 256; i++) {
		c = i;
		for (int j = 0; j < 8; j++) c = c & 1 ? 0xEDB88320 ^ (c >> 1) : (c >> 1);
		table[i] = c;
	}
	for (const char *p = s; p < s + strlen(s); p++)
		crc = (crc >> 8) ^ table[crc & 0xff ^ *p];
	return ~crc;
}


int main(int argc, char *argv[]) {
	if (argc == 2)
		return printf("%lx\n", crc32(argv[1]));
	else
		return 1;
}
